File: shared/schemas.py
"""Data contracts: exact signed shapes. Unknown extra fields are REJECTED,
not ignored — a signer and verifier can never silently disagree.
"""
from shared.errors import SchemaError

CRED_SIGNED_FIELDS = frozenset({"v", "iss", "sub", "vid", "a", "r", "iat", "exp", "cnf"})
CRED_REQUIRED_FIELDS = CRED_SIGNED_FIELDS | {"s"}

BUNDLE_SIGNED_FIELDS = frozenset(
    {"v", "iss", "vid", "pub", "next_pub", "revoked", "minors", "iat", "exp"})
BUNDLE_REQUIRED_FIELDS = BUNDLE_SIGNED_FIELDS | {"s"}

CHALLENGE_FIELDS = frozenset({"n", "vid", "exp"})
PROOF_FIELDS = frozenset({"n", "ts", "sig"})

CRED_MAX_BYTES = 2048  # full {c,p} presentation payload cap


def _is_plain_int(value: object) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def signed_body(cred: dict) -> dict:
    """The exact dict covered by the issuer signature. Raises SchemaError if
    the credential carries fields outside the contract."""
    unknown = set(cred.keys()) - CRED_REQUIRED_FIELDS
    if unknown:
        raise SchemaError(f"unknown credential fields: {sorted(unknown)}")
    return {k: cred[k] for k in CRED_SIGNED_FIELDS}


def validate_credential(cred: object) -> str | None:
    """Shape + type validation. Returns None if OK, else a reason code."""
    if not isinstance(cred, dict):
        return "MALFORMED"
    if not CRED_REQUIRED_FIELDS.issubset(cred.keys()):
        return "MALFORMED"
    try:
        signed_body(cred)
    except SchemaError:
        return "MALFORMED"
    if not _is_plain_int(cred.get("exp")) or not _is_plain_int(cred.get("iat")):
        return "MALFORMED"
    for key in ("sub", "vid", "a", "iss", "cnf"):
        if not isinstance(cred.get(key), str):
            return "MALFORMED"
    if not _is_plain_int(cred.get("r")) or cred["r"] not in (0, 1):
        return "MALFORMED"
    if not _is_plain_int(cred.get("v")):
        return "MALFORMED"
    return None


def validate_proof(proof: object) -> str | None:
    """Shape + type validation for a holder proof. None if OK, else reason."""
    if not isinstance(proof, dict):
        return "MALFORMED"
    if set(proof.keys()) != PROOF_FIELDS:
        return "MALFORMED"
    if not isinstance(proof.get("n"), str):
        return "MALFORMED"
    if not _is_plain_int(proof.get("ts")):
        return "MALFORMED"
    if not isinstance(proof.get("sig"), str):
        return "MALFORMED"
    return None


def validate_challenge(ch: object) -> str | None:
    """Shape + type validation for a shop challenge. None if OK, else reason."""
    if not isinstance(ch, dict):
        return "MALFORMED"
    if set(ch.keys()) != CHALLENGE_FIELDS:
        return "MALFORMED"
    if not isinstance(ch.get("n"), str):
        return "MALFORMED"
    if not isinstance(ch.get("vid"), str):
        return "MALFORMED"
    if not _is_plain_int(ch.get("exp")):
        return "MALFORMED"
    return None


def validate_bundle_shape(bundle: object) -> str | None:
    """Shape + type validation for a trust bundle (signature checked separately)."""
    if not isinstance(bundle, dict):
        return "MALFORMED"
    if not BUNDLE_REQUIRED_FIELDS.issubset(bundle.keys()):
        return "MALFORMED"
    unknown = set(bundle.keys()) - BUNDLE_REQUIRED_FIELDS
    if unknown:
        return "MALFORMED"
    if not isinstance(bundle.get("pub"), str) or len(bundle["pub"]) != 64:
        return "MALFORMED"
    nxt = bundle.get("next_pub")
    if nxt is not None and (not isinstance(nxt, str) or len(nxt) != 64):
        return "MALFORMED"
    if not _is_plain_int(bundle.get("v")):
        return "MALFORMED"
    if not _is_plain_int(bundle.get("iat")) or not _is_plain_int(bundle.get("exp")):
        return "MALFORMED"
    if not isinstance(bundle.get("revoked"), list):
        return "MALFORMED"
    if not isinstance(bundle.get("minors"), list):
        return "MALFORMED"
    if not isinstance(bundle.get("iss"), str) or not isinstance(bundle.get("vid"), str):
        return "MALFORMED"
    if not isinstance(bundle.get("s"), str):
        return "MALFORMED"
    return None


def bundle_sig_body(bundle: dict) -> dict:
    """The exact dict the issuer signs (everything except "s")."""
    return {k: bundle[k] for k in BUNDLE_SIGNED_FIELDS}

File: vercel/lib/schemas.ts
// Zod schemas - port of shared/schemas.py
// Exact signed shapes. Unknown extra fields are REJECTED, not ignored.

import { z } from 'zod';

// ============================================================================
// Constants
// ============================================================================

export const CRED_SIGNED_FIELDS = [
  'v', 'iss', 'sub', 'vid', 'a', 'r', 'iat', 'exp', 'cnf', 'did'
] as const;

export const CRED_REQUIRED_FIELDS = [...CRED_SIGNED_FIELDS, 's'];

export const BUNDLE_SIGNED_FIELDS = [
  'v', 'iss', 'vid', 'pub', 'next_pub', 'revoked', 'minors', 'iat', 'exp'
] as const;

export const BUNDLE_REQUIRED_FIELDS = [...BUNDLE_SIGNED_FIELDS, 's'];

export const CHALLENGE_FIELDS = ['n', 'vid', 'exp'] as const;
export const PROOF_FIELDS = ['n', 'ts', 'sig'] as const;

export const CRED_MAX_BYTES = 2048;

// ============================================================================
// Type helpers
// ============================================================================

function plainInt(): z.ZodNumber | z.ZodEffects<z.ZodNumber, number, number> {
  return z.number().int().refine(n => !Number.isNaN(n), 'must be integer');
}

function hexString(len?: number): z.ZodString {
  const schema = z.string().regex(/^[0-9a-f]+$/i, 'must be hex');
  if (len) return schema.length(len);
  return schema;
}

function b64uString(): z.ZodString {
  return z.string().regex(/^[A-Za-z0-9_-]+$/, 'must be base64url');
}

// ============================================================================
// Credential schemas
// ============================================================================

export const CredentialSignedSchema = z.object({
  v: plainInt(),
  iss: z.string().min(1).max(64),
  sub: hexString(32),
  vid: z.string().min(1).max(32),
  a: z.literal('over_18'),
  r: z.union([z.literal(0), z.literal(1)]),
  iat: plainInt(),
  exp: plainInt(),
  cnf: b64uString(),
  did: hexString(32)
}).strict();

export const CredentialSchema = CredentialSignedSchema.extend({
  s: b64uString()
}).strict();

export type Credential = z.infer<typeof CredentialSchema>;
export type CredentialSigned = z.infer<typeof CredentialSignedSchema>;

// ============================================================================
// Proof schema
// ============================================================================

export const ProofSchema = z.object({
  n: z.string().min(1).max(64),
  ts: plainInt(),
  sig: b64uString()
}).strict();

export type Proof = z.infer<typeof ProofSchema>;

// ============================================================================
// Challenge schema
// ============================================================================

export const ChallengeSchema = z.object({
  n: z.string().length(32).regex(/^[0-9a-f]+$/, 'nonce must be 32-char hex'),
  vid: z.string().min(1).max(32),
  exp: plainInt()
}).strict();

export type Challenge = z.infer<typeof ChallengeSchema>;

// ============================================================================
// Bundle schemas
// ============================================================================

export const BundleSignedSchema = z.object({
  v: plainInt(),
  iss: z.string().min(1).max(64),
  vid: z.string().min(1).max(32),
  pub: hexString(64),
  next_pub: z.union([hexString(64), z.null()]),
  revoked: z.array(hexString(32)),
  minors: z.array(hexString(32)),
  iat: plainInt(),
  exp: plainInt()
}).strict();

export const BundleSchema = BundleSignedSchema.extend({
  s: b64uString()
}).strict();

export type Bundle = z.infer<typeof BundleSchema>;
export type BundleSigned = z.infer<typeof BundleSignedSchema>;

// ============================================================================
// Verification request
// ============================================================================

export const VerifyRequestSchema = z.object({
  c: CredentialSchema,
  p: ProofSchema
}).strict();

export type VerifyRequest = z.infer<typeof VerifyRequestSchema>;

// ============================================================================
// OTP verification request
// ============================================================================

export const VerifyCodeRequestSchema = z.object({
  code: z.string().length(6).regex(/^\d{6}$/, 'code must be 6 digits')
}).strict();

export type VerifyCodeRequest = z.infer<typeof VerifyCodeRequestSchema>;

// ============================================================================
// Admin schemas
// ============================================================================

export const EnrollmentCodeRequestSchema = z.object({
  user_id: z.string().min(1).max(32)
}).strict();

export const RevokeRequestSchema = z.object({
  user_id: z.string().min(1).max(32)
}).strict();

export const RotateRequestSchema = z.object({
  activate: z.boolean().default(false)
}).strict();

export const SyncRequestSchema = z.object({
  bundle: BundleSchema
}).strict();

export const OTPSecretsRequestSchema = z.object({
  secrets: z.record(z.string().length(32), z.string().length(64))
}).strict();

// ============================================================================
// Validation helpers (return error code or null)
// ============================================================================

export const REASONS = [
  'NO_TRUSTBUNDLE', 'MALFORMED', 'TOO_LARGE', 'WRONG_VERIFIER',
  'ISSUER_MISMATCH', 'EXPIRED', 'BADSIG', 'BAD_PROOF', 'UNKNOWN_CHALLENGE',
  'REVOKED', 'NOT_ADULT', 'OK', 'OK_OTP', 'BAD_OTP',
  'ENROLL_FAILED', 'REVOKED_USER', 'BAD_PUBKEY', 'UNKNOWN_CODE',
  'CODE_USED', 'CODE_EXPIRED', 'ROLLBACK', 'OTP_DISABLED', 'UNAUTHORIZED'
] as const;

export type ReasonCode = typeof REASONS[number];

export function checkReason(reason: string): reason is ReasonCode {
  return REASONS.includes(reason as ReasonCode);
}

function isPlainInt(n: unknown): n is number {
  return Number.isInteger(n) && !Number.isNaN(n);
}

export function validateCredential(cred: unknown): ReasonCode | null {
  const result = CredentialSchema.safeParse(cred);
  if (!result.success) return 'MALFORMED';
  return null;
}

export function validateProof(proof: unknown): ReasonCode | null {
  const result = ProofSchema.safeParse(proof);
  if (!result.success) return 'MALFORMED';
  return null;
}

export function validateChallenge(ch: unknown): ReasonCode | null {
  const result = ChallengeSchema.safeParse(ch);
  if (!result.success) return 'MALFORMED';
  return null;
}

export function validateBundleShape(bundle: unknown): ReasonCode | null {
  const result = BundleSchema.safeParse(bundle);
  if (!result.success) return 'MALFORMED';
  return null;
}

// ============================================================================
// Error response helper
// ============================================================================

export function err(code: ReasonCode, detail?: string): Response {
  const body = detail ? { error: code, detail } : { error: code };
  const status = code === 'MALFORMED' ? 400 :
                 code === 'UNKNOWN_CODE' || code === 'CODE_USED' || code === 'CODE_EXPIRED' || code === 'BAD_PUBKEY' || code === 'ENROLL_FAILED' ? 400 :
                 code === 'REVOKED_USER' ? 403 :
                 code === 'UNAUTHORIZED' ? 401 :
                 code === 'BADSIG' || code === 'ROLLBACK' || code === 'ISSUER_MISMATCH' ? 409 :
                 code === 'OTP_DISABLED' ? 404 : 500;
  return Response.json(body, { status });
}